packages/utils/src/modules-sdk/internal-module-service-factory.ts

```diff
@@ -0,0 +1,427 @@
+          MedusaError.Types.NOT_FOUND,
+          `${
+            primaryKeys.length === 1
+              ? `${lowerCaseFirst(model.name) + " - " + primaryKeys[0]}`
+              : `${lowerCaseFirst(model.name)} - ${primaryKeys.join(", ")}`
+          } must be defined`
+        )
+      }
+
+      let primaryKeysCriteria = {}
+      if (primaryKeys.length === 1) {
+        primaryKeysCriteria[primaryKeys[0]] = idOrObject
+      } else {
+        const idOrObject_ = Array.isArray(idOrObject)
+          ? idOrObject
+          : [idOrObject]
+        primaryKeysCriteria = idOrObject_.map((primaryKeyValue) => ({
+          $and: primaryKeys.map((key) => ({ [key]: primaryKeyValue[key] })),
+        }))
+      }
+
+      const queryOptions = buildQuery(primaryKeysCriteria, config)
+
+      const entities = await this[propertyRepositoryName].find(
+        queryOptions,
+        sharedContext
+      )
+
+      if (!entities?.length) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_FOUND,
+          `${model.name} with ${primaryKeys.join(", ")}: ${
+            Array.isArray(idOrObject)
+              ? idOrObject.map((v) =>
+                  [isString(v) ? v : Object.values(v)].join(", ")
+                )
+              : idOrObject
+          } was not found`
+        )
+      }
+
+      return entities[0]
+    }
+
+    @InjectManager(propertyRepositoryName)
+    async list(
+      filters: FilterQuery<any> | BaseFilterable<FilterQuery<any>> = {},
+      config: FindConfig<any> = {},
+      @MedusaContext() sharedContext: Context = {}
+    ): Promise<TEntity[]> {
+      const queryOptions = buildQuery(filters, config)
+
+      return await this[propertyRepositoryName].find(
+        queryOptions,
+        sharedContext
+      )
+    }
+
+    @InjectManager(propertyRepositoryName)
+    async listAndCount(
+      filters: FilterQuery<any> | BaseFilterable<FilterQuery<any>> = {},
+      config: FindConfig<any> = {},
+      @MedusaContext() sharedContext: Context = {}
+    ): Promise<[TEntity[], number]> {
+      const queryOptions = buildQuery(filters, config)
+
+      return await this[propertyRepositoryName].findAndCount(
+        queryOptions,
+        sharedContext
+      )
+    }
+
+    create(data: any, sharedContext?: Context): Promise<TEntity>
+    create(data: any[], sharedContext?: Context): Promise<TEntity[]>
+
+    @InjectTransactionManager(shouldForceTransaction, propertyRepositoryName)
+    async create(
+      data: any | any[],
+      @MedusaContext() sharedContext: Context = {}
+    ): Promise<TEntity | TEntity[]> {
+      const data_ = Array.isArray(data) ? data : [data]
+      const entities = await this[propertyRepositoryName].create(
+        data_,
+        sharedContext
+      )
+
+      return Array.isArray(data) ? entities : entities[0]
+    }
+
+    update(data: any[], sharedContext?: Context): Promise<TEntity[]>
+    update(data: any, sharedContext?: Context): Promise<TEntity>
```

Comment:
suggestion: update should only work with selector + data. If you want to update with an entity you can use upsert. Wdyt?

Reply:
Thats true, we can rework that 👌

Reply:
maybe we can do that in a follow-up pr just to merge this one asap? wdyt?